rtcp-packet.go

```diff
@@ -0,0 +1,293 @@
+}
+
+func (p *RTCPCompoundPacket) GetSRTCPIndex() uint32 {
+  esrtcpWord := binary.BigEndian.Uint32(p.GetESRTCPWord())
+
+  // This AND sets the first bit that is the E-bit to 0
+  return esrtcpWord & uint32(0x7fffffff)
+}
+
+func (p *RTCPCompoundPacket) GetE() bool {
+  return bool((p.GetESRTCPWord()[0] & byte(128)) == 128)
+}
+
+func (p *RTCPCompoundPacket) getAAD() []byte {
+  srtcpIndexLine := make([]byte, len(p.GetESRTCPWord()))
+  copy(srtcpIndexLine, p.GetESRTCPWord())
+
+  headerBuffer := make([]byte, len(p.header.buffer))
+  copy(headerBuffer, p.header.buffer)
+
+  return append(headerBuffer, p.GetESRTCPWord()...)
+}
+
+// https://tools.ietf.org/html/rfc7714#section-9.1
+/*
+     0  1  2  3  4  5  6  7  8  9 10 11
+   +--+--+--+--+--+--+--+--+--+--+--+--+
+   |00|00|    SSRC   |00|00|0+SRTCP Idx|---+
+   +--+--+--+--+--+--+--+--+--+--+--+--+   |
+                                           |
+   +--+--+--+--+--+--+--+--+--+--+--+--+   |
+   |         Encryption Salt           |->(+)
+   +--+--+--+--+--+--+--+--+--+--+--+--+   |
+                                           |
+   +--+--+--+--+--+--+--+--+--+--+--+--+   |
+   |       Initialization Vector       |<--+
+   +--+--+--+--+--+--+--+--+--+--+--+--+
+*/
+
+func (p* RTCPCompoundPacket) gcmIV(salt []byte) []byte {
+  srtcpIndex := make([]byte, 4)
+  binary.BigEndian.PutUint32(srtcpIndex, p.GetSRTCPIndex())
+
+  ssrc := make([]byte, 4)
+  binary.BigEndian.PutUint32(ssrc, p.header.GetSenderSSRC())
+
+  iv := make([]byte, 12)
+
+  iv[2] = ssrc[0]
+  iv[3] = ssrc[1]
+  iv[4] = ssrc[2]
+  iv[5] = ssrc[3]
+
+  iv[8] = srtcpIndex[0]
+  iv[9] = srtcpIndex[1]
+  iv[10] = srtcpIndex[2]
+  iv[11] = srtcpIndex[3]
+
+  for i := range iv {
+    iv[i] ^= salt[i]
+  }
+
+  return iv
+}
+
+func (p *RTCPCompoundPacket) DecryptGCM(key, salt []byte) error {
+  if !p.GetE() {
+    return errors.New("srtcp: Encryption flag not set")
+  }
+
+  block, err := aes.NewCipher(key)
+  if err != nil {
+    return err
+  }
+
+  gcm, err := cipher.NewGCM(block)
+  if err != nil {
+    return err
+  }
+
+  iv := p.gcmIV(salt)
+
+  aad := p.getAAD()
+  ct := p.buffer
+
+  _, err = gcm.Open(p.buffer[0:0], iv, ct, aad)
+  if err != nil {
+    return err
+  }
+
+  p.buffer = p.buffer[:len(p.buffer)-gcm.Overhead()]
+
+  return nil
+}
+
+func (p *RTCPCompoundPacket) EncryptGCM(key, salt []byte) error {
+  block, err := aes.NewCipher(key)
+  if err != nil {
+    return err
+  }
+
+  gcm, err := cipher.NewGCM(block)
+  if err != nil {
+    return err
+  }
+
+  iv := p.gcmIV(salt)
+
+  tag := make([]byte, gcm.Overhead())
+  p.buffer = append(p.buffer, tag...)
+
+  aad := p.getAAD()
+  pt := p.buffer[:len(p.buffer)-len(tag)]
+
+  gcm.Seal(p.buffer[0:0], iv, pt, aad)
+
+  return nil
+}
+
+func (p *RTCPCompoundPacket) GetBuffer() []byte {
+  buffer := append(p.header.buffer, p.buffer...)
+  buffer = append(buffer, p.appendix...)
+  return buffer
+}
+
+func NewSRTCPPacket(buffer []byte) (*RTCPCompoundPacket, error) {
```

Comment:
I would just inline this into `RTPSession.DecodeRTCP()`, as is done for SRTP.

Reply:
I am also using this in the testcases and I do not want to call `DecodeRTCP` from the testcases at this point

Reply:
If you only need it in the test-cases, put in the `_test.go` file.

Reply:
Then I would have the same code in DecodeSRTCP and a testcase. In this way, I also have test coverage on the NewSRTCPPacket function